### PR TITLE
refactor: extract Mailbox type (#3126)

### PR DESCRIPTION
**Pull request by tustvold**
This extracts the delete mailbox functionality into its own type, with the addition of synchronisation logic to allow coordination between persist lifecycle actions and the background worker.

An argument could be made that this additional synchronisation is redundant and could just make use of the single catalog transaction, but this would require duplicating logic in multiple places or pushing the mailbox into the PreservedCatalog. Having discussed with @crepererum I think this is simpler

### CODE REVIEW

**Review by alamb — APPROVED** [review]
I think the code does what it says , though I think it is worth considering using a tokio `Channel`

**Review by crepererum — APPROVED** [review]

File: internal_types/src/mailbox.rs [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    }
+}
+
+impl<T> Mailbox<T> {
+    /// Add an item to the inbox
+    pub fn push(&self, item: T) {
+        self.inbox.lock().push(item)
+    }
+
+    /// Get a handle to consumer data from this [`Mailbox`]
```
[the remark this answers is not in the document]
Comment (tustvold) [reply internal_types/src/mailbox.rs:34]: I too like to consumer data :laughing:

**Comment by tustvold** [comment]
> I think it is worth considering using a tokio Channel

I don't think this is a comparable abstraction to a channel:

* The contents can be inspected prior to flushing (important as catalog transactions are fallible)
* Potentially multiple consumers acquire exclusive access for the duration of a catalog transaction
* We need to be sure to take everything in the mailbox and not just some stuff

If there is some way to get this some other way, I'd be interested to know :smile:

File: server/src/db.rs [hunk 2]
```diff
@@ -839,27 +840,15 @@ impl Db {
...
-                            // https://github.com/influxdata/influxdb_iox/issues/2626 is implemented
-                            guard.drain(0..todo.len());
-                        }
-                        Err(e) => {
-                            error!(%e, "cannot preserve delete predicates");
-                        }
-                    }
+                let handle = self.delete_predicates_mailbox.consume().await;
+                match self.preserve_delete_predicates(handle.outbox()).await {
+                    Ok(()) => handle.flush(),
```
Comment (alamb) [inline server/src/db.rs:845]: should we possibly "unflush" the handle here or `panic!`? It seems like as written the deletes are dropped (not successfully applied)

**Comment by alamb** [comment]
https://github.com/influxdata/influxdb_iox/pull/3148 <-- added the difference between channels 👍

File: server/src/db.rs (hunk 2 above)
Comment (tustvold) [reply server/src/db.rs:845]: I'm not sure I follow, if we get an error the handle is dropped and nothing is removed from the collection?

Comment (alamb) [reply server/src/db.rs:845]: Ah -- I see -- because the handle holds the lock on the outbox. Sorry my bad.

**Pull request merged**